### ISSUE TITLE
Raise Python exception before encountering Fortran STOP

### ISSUE THREAD

**Issue by MarkWieczorek** [SHTOOLS/SHTOOLS#25]
Summary
---------
The Fortran 95 SHTOOLS routines make use of the statement `STOP` in order to stop execution when a problem occurs. However, when called from within Python, the Fortran `STOP` statement will not only exit the Fortran program, but it will stop the Python kernel.

There is a potential way to get around this using Python callback functions. In essence, in the Fortran routine, you call an undefined function just before encountering the `STOP`, you define this function with `intent(callback,hide)` in the Python signature file, you define a function in Python whose sole purpose is to raise an exception, and then you set the function name in the module equal to the Python function that raises the exception.

This works, but only when the Fortran function is called directly from Python, and when the Python callback function is in this Fortran routine. However, this does not appear to work if another Fortran routine calls the Fortran routine that contains the callback function. This is detailed here in this issue: https://github.com/numpy/numpy/issues/7855

This is a major problem for SHTOOLS, because each Python function first calls a generic Fortran wrapper function that then calls the real Fortran routine.

Solutions
---------
Solution 1. This is most likely a bug in numpy/f2py. Wait for them to fix this. I have opened an issue there, but I am not optimistic for a response. 

Solution 2. Add an optional variable to each SHTOOLS function called `ExitStatus`. Set this variable equal to 0 when no errors occur, and something other than 0 when an error occurs. After each call to a SHTOOLS routine in the Fortran routines, it will be necessary to check if this value if non-zero, and if so, to return this value immediately to the calling function. Eventually, you will end in the Fortran Python wrapper function, the wrapper function checks if `ExitStatus` is non-zero, and if so, this routine raises a Python exception from a callback function.

Solution 2 will require editing of all SHTOOLS source files. Furthermore, after each call to an SHTOOLS routine, it will be necessary to check the value of `ExitStatus`. If this check is in a deep do loop, this might affect performance of the routine. Also, it is not clear how to set an exit status for a Fortran program (the returned value could be set equal to a predefined value, but this is not an optimal solution).

See also: https://github.com/SHTOOLS/SHTOOLS/issues/9

**Comment by andreww** [comment]
I think I can explain why the exception is not working in a helpful way for `routine1` in numpy/numpy#7855, but I don't think I can think of anything better than solution 2.

My mental model of python's exception handling (no doubt vastly simplified or wrong) is that there is a flag in the interpreter that is set when an exception is raised. This is checked by the interpreter each time a *python* function is called, and if the flag is set instead of the python function being called the call stack is examined and the exception handling machinery activated. To operate, the python interpreter calls a lots of different *interpreter* functions (written in C). Some of these functions will do the job of checking and setting python's exception flag. The C functions have their own call stack and frame pointer which works in the same way as for any other compiled code. An interpreter function calling another interpreter function does not result in the exception flag being examined. When f2py is used to compile a Fortran subroutine you end up with something that looks like a python function and a function that behaves like an interpreter function. When you call `routine1` from python, it calls `routine2` as an "interpreter" function, this calls `f` and sets the exception flag. However, the the state of that flag is not checked on return from `routine2` (this is just one fortran subroutine calling another, it does not know that it is running inside the process running the python interpreter) but only when you get back to something that looks like a "python function" (`routine1`). One thing to note is that even for the `routine1` call, by the time you get back to the python interpreter you do end up with an exception (no more python gets run and the exception is raised) it's just that all of `routine2` has been run by this point. The callback machinery seems to know about the exception (I guess this is where the "Call-back cb_pystop_in_routine2__user__routines failed." line comes from) but does not seem to want to do anything about it. I don't know if it would be possible to fiddle with the interpreter call stack at this point to skip the rest of `routine2`, I suspect not. I've tried fiddling around with the callback function (e.g. making it call another python function which raises the exception, having a return value) but nothing seems to stop the rest of `routine2` being run.

So, I think solution 2 is probably the best bet. The other advantage of this is that it would be possible to avoid `stop` in Fortran if people want to run SHTOOLS inside of MPI (`mpi_abort()` is probably better) or want to arrange for a core dump if they pass duff data into an SHTOOLS routine. We've got some fairly fancy (overly baroque?) implementations of exception handling in Fortran inside of [FoX](https://github.com/andreww/fox) but these are probably overkill and are not designed for performance critical code.
